File: server/src/routes/favorites.ts
import { Router, Request, Response } from 'express';
import { prisma } from '../db';
import { authRequired } from '../middleware/auth';

const router = Router();

router.get('/', authRequired, async (req: Request, res: Response) => {
  const favs = await prisma.favorite.findMany({
    where: { userId: req.user!.userId },
    include: { location: true },
    orderBy: { createdAt: 'desc' }
  });
  const parsed = favs.map(f => ({
    ...f,
    location: { ...f.location, tags: JSON.parse(f.location.tags), images: JSON.parse(f.location.images) }
  }));
  res.json(parsed);
});

router.post('/:locationId', authRequired, async (req: Request, res: Response) => {
  const locationId = parseInt(req.params.locationId);
  const existing = await prisma.favorite.findFirst({
    where: { userId: req.user!.userId, locationId }
  });
  if (existing) {
    await prisma.favorite.delete({ where: { id: existing.id } });
    return res.json({ favorited: false });
  }
  await prisma.favorite.create({
    data: { userId: req.user!.userId, locationId }
  });
  res.json({ favorited: true });
});

export default router;

File: server/routes/favorites.py
from flask import Blueprint, request, jsonify
from models import db, Favorite, Location
from utils.auth import auth_required

favorites_bp = Blueprint('favorites', __name__)


@favorites_bp.route('/', methods=['GET'])
@auth_required
def list_favorites():
    favs = (
        db.session.query(Favorite, Location)
        .join(Location, Favorite.location_id == Location.id)
        .filter(Favorite.user_id == request.user_id)
        .order_by(Favorite.created_at.desc())
        .all()
    )
    result = []
    for fav, loc in favs:
        item = fav.__dict__.copy()
        item.pop('_sa_instance_state', None)
        item['location'] = loc.to_dict()
        result.append(item)
    return jsonify(result)


@favorites_bp.route('/<int:location_id>', methods=['POST'])
@auth_required
def toggle_favorite(location_id):
    existing = Favorite.query.filter_by(user_id=request.user_id, location_id=location_id).first()
    if existing:
        db.session.delete(existing)
        db.session.commit()
        return jsonify({'favorited': False})
    fav = Favorite(user_id=request.user_id, location_id=location_id)
    db.session.add(fav)
    db.session.commit()
    return jsonify({'favorited': True})
